File: functionapi-api/src/main/java/io/github/cottonmc/functionapi/documentation/MarkdownPrinter.java
package io.github.cottonmc.functionapi.documentation;

import java.io.FileNotFoundException;
import java.io.PrintWriter;
import java.util.List;

/**
 * Helper to print markdown text. Used for documentation!
 *
 * */
public class MarkdownPrinter {
    private final PrintWriter printWriter;

    private final String separator = System.lineSeparator();

    public MarkdownPrinter(String filename) throws FileNotFoundException {
        printWriter = new PrintWriter(filename);
    }

    public MarkdownPrinter header1(String text){
        addHeader(text,"-");
        return this;
    }

    public MarkdownPrinter header2(String text){
        addHeader(text,"=");
        return this;
    }

    public MarkdownPrinter paragraph(String text){
        printWriter.println(text+separator);
        return this;
    }


    public MarkdownPrinter bold(String text){
        formatText(text,"**");
        return this;
    }

    public MarkdownPrinter italics(String text){
        formatText(text,"*");
        return this;
    }

    public MarkdownPrinter boldItalics(String text){
        formatText(text,"***");
        return this;
    }

    public MarkdownPrinter orderedList(List<String> text){
        int index = 1;
        for (String s : text) {
            printWriter.println(index+". "+text);
        }
        return this;
    }
    public MarkdownPrinter unOrderedList(List<String> text){
        for (String s : text) {
            printWriter.println(" - "+text);
        }
        return this;
    }

    public MarkdownPrinter table(List<List<String>> dataWithHeader){
        List<String> header = dataWithHeader.get(0);
        /*| Syntax | Description |
| ----------- | ----------- |
| Header | Title |
| Paragraph | Text |*/
        printWriter.print("|");
        for (String s : header) {
            printWriter.print(s+"|");
        }
        printWriter.println();
        printWriter.print("|");
        for (String s : header) {
            int length = s.length();
            for (int i = 0; i < length; i++) {
                printWriter.print("-");
            }
            printWriter.print("|");
        }
        printWriter.println();

        for (int i = 1; i <dataWithHeader.size(); i++) {
            List<String> data = dataWithHeader.get(i);
            printWriter.print("|");
            for (String s : data) {
                printWriter.print(s+"|");
            }
            printWriter.println();
        }

        return this;
    }

    private void formatText(String text,String format){
        printWriter.print(format+text+format);
    }

    private void addHeader(String text,String marker){
        int length = text.length();
        StringBuilder header = new StringBuilder(text + System.lineSeparator());

        for (int i = 0; i < length; i++) {
            header.append(marker);
        }

        printWriter.println(header);

    }

    public void finish(){
        printWriter.flush();
    }

}

File: functionapi-content/src/main/java/io/github/cottonmc/functionapi/content/documentation/MarkdownPrinter.kt
package io.github.cottonmc.functionapi.content.documentation

import java.io.PrintWriter

/**
 * Helper to print markdown text. Used for documentation!
 *
 */
class MarkdownPrinter(filename: String?) {
    private val printWriter: PrintWriter
    private val separator = System.lineSeparator()
    fun header1(text: String): MarkdownPrinter {
        addHeader(text, "-")
        return this
    }

    fun header2(text: String): MarkdownPrinter {
        addHeader(text, "=")
        return this
    }

    fun paragraph(text: String): MarkdownPrinter {
        printWriter.println(text + separator)
        return this
    }

    fun bold(text: String): MarkdownPrinter {
        formatText(text, "**")
        return this
    }

    fun italics(text: String): MarkdownPrinter {
        formatText(text, "*")
        return this
    }

    fun boldItalics(text: String): MarkdownPrinter {
        formatText(text, "***")
        return this
    }

    fun orderedList(text: List<String?>): MarkdownPrinter {
        val index = 1
        for (s in text) {
            printWriter.println("$index. $text")
        }
        return this
    }

    fun unOrderedList(text: List<String?>): MarkdownPrinter {
        for (s in text) {
            printWriter.println(" - $text")
        }
        return this
    }

    fun table(dataWithHeader: List<List<String>>): MarkdownPrinter {
        val header = dataWithHeader[0]
        /*| Syntax | Description |
| ----------- | ----------- |
| Header | Title |
| Paragraph | Text |*/printWriter.print("|")
        for (s in header) {
            printWriter.print("$s|")
        }
        printWriter.println()
        printWriter.print("|")
        for (s in header) {
            val length = s.length
            for (i in 0 until length) {
                printWriter.print("-")
            }
            printWriter.print("|")
        }
        printWriter.println()
        for (i in 1 until dataWithHeader.size) {
            val data = dataWithHeader[i]
            printWriter.print("|")
            for (s in data) {
                printWriter.print("$s|")
            }
            printWriter.println()
        }
        return this
    }

    private fun formatText(text: String, format: String) {
        printWriter.print(format + text + format)
    }

    private fun addHeader(text: String, marker: String) {
        val length = text.length
        val header = StringBuilder(text + System.lineSeparator())
        for (i in 0 until length) {
            header.append(marker)
        }
        printWriter.println(header)
    }

    fun finish() {
        printWriter.flush()
    }

    init {
        printWriter = PrintWriter(filename)
    }
}